pkg/tso/allocator_manager.go

```diff
@@ -169,38 +189,127 @@ type AllocatorManager struct {
 
 // NewAllocatorManager creates a new TSO Allocator Manager.
 func NewAllocatorManager(
-	m Member,
+	ctx context.Context,
+	startGlobalLeaderLoop bool,
+	keyspaceGroupID uint32,
+	member ElectionMember,
 	rootPath string,
 	storage endpoint.TSOStorage,
 	enableLocalTSO bool,
 	saveInterval time.Duration,
 	updatePhysicalInterval time.Duration,
+	leaderLease int64,
 	tlsConfig *grpcutil.TLSConfig,
 	maxResetTSGap func() time.Duration,
 ) *AllocatorManager {
+	ctx, cancel := context.WithCancel(ctx)
 	allocatorManager := &AllocatorManager{
-		enableLocalTSO:         enableLocalTSO,
-		member:                 m,
+		ctx:                    ctx,
+		cancel:                 cancel,
+		ksgID:                  keyspaceGroupID,
+		member:                 member,
 		rootPath:               rootPath,
 		storage:                storage,
+		enableLocalTSO:         enableLocalTSO,
 		saveInterval:           saveInterval,
 		updatePhysicalInterval: updatePhysicalInterval,
+		leaderLease:            leaderLease,
 		maxResetTSGap:          maxResetTSGap,
 		securityConfig:         tlsConfig,
 	}
 	allocatorManager.mu.allocatorGroups = make(map[string]*allocatorGroup)
 	allocatorManager.mu.clusterDCLocations = make(map[string]*DCLocationInfo)
 	allocatorManager.localAllocatorConn.clientConns = make(map[string]*grpc.ClientConn)
+
+	// Set up the Global TSO Allocator here, it will be initialized once the member campaigns leader successfully.
+	allocatorManager.setUpGlobalAllocator(startGlobalLeaderLoop)
+
 	return allocatorManager
 }
 
+// setUpGlobalAllocator is used to set up the global allocator, which will initialize the allocator and put it into
+// an allocator daemon. An TSO Allocator should only be set once, and may be initialized and reset multiple times
+// depending on the election.
+func (am *AllocatorManager) setUpGlobalAllocator(startGlobalLeaderLoop bool) {
+	am.mu.Lock()
+	defer am.mu.Unlock()
+
+	allocator := NewGlobalTSOAllocator(am.ctx, am, startGlobalLeaderLoop)
+	// Create a new allocatorGroup
+	ctx, cancel := context.WithCancel(am.ctx)
+	am.mu.allocatorGroups[GlobalDCLocation] = &allocatorGroup{
+		dcLocation: GlobalDCLocation,
+		ctx:        ctx,
+		cancel:     cancel,
+		leadership: am.member.GetLeadership(),
+		allocator:  allocator,
+	}
+
+	if startGlobalLeaderLoop {
+		am.wg.Add(1)
+		go am.tsoAllocatorLoop()
+	}
+}
+
+// tsoAllocatorLoop is used to run the TSO Allocator updating daemon.
+// tso service starts the loop here, but pd starts its own loop.
+func (am *AllocatorManager) tsoAllocatorLoop() {
+	defer logutil.LogPanic()
+	defer am.wg.Done()
+
+	ctx, cancel := context.WithCancel(am.ctx)
+	defer cancel()
```

Comment:
Perhaps, we don't need the context here.

Reply:
removed line 260-261